File: backend/promptforge/schema.py
"""Typed contract for an evaluation (PRD section 5 JSON, minus overall_score).

`overall_score` is intentionally NOT a model output - it is computed deterministically
in `scoring.py` so the band (and the required-dimension cap) never depends on the
model's arithmetic. This module is the reusable contract for the future web backend.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

Score = Literal["Poor", "Needs Work", "Good", "Excellent"]


class Dimension(BaseModel):
    score: Score = Field(description="Poor / Needs Work / Good / Excellent")
    assessment: str = Field(description="1-2 sentences explaining the score")
    advice: str = Field(description="one specific, actionable improvement")


class Scorecard(BaseModel):
    clarity: Dimension
    guidelines: Dimension
    structure: Dimension
    examples: Dimension


class Evaluation(BaseModel):
    prompt_evaluated: str = Field(description="the user's original prompt, verbatim")
    scorecard: Scorecard
    priority_fix: str = Field(
        description="the single most impactful change to make first"
    )
    revised_prompt: str = Field(description="the full rewritten prompt, ready to use")
    suggested_category: str = Field(description="a single short category for filing")
    suggested_tags: list[str] = Field(
        default_factory=list, description="2-5 short lowercase tags for search"
    )

File: frontend/lib/schema.ts
// Typed contract for an evaluation, mirroring promptforge/schema.py (Python CLI).
// Uses `zod/v4` so the schema is the exact module @anthropic-ai/sdk's zodOutputFormat
// expects. `overall_score` is computed server-side (scoring.ts), never from the model.
import * as z from "zod/v4";

export const SCORES = ["Poor", "Needs Work", "Good", "Excellent"] as const;
export const ScoreSchema = z.enum(SCORES);
export type Score = (typeof SCORES)[number];

export const DimensionSchema = z.object({
  score: ScoreSchema.describe("Poor / Needs Work / Good / Excellent"),
  assessment: z.string().describe("1-2 sentences explaining the score"),
  advice: z.string().describe("one specific, actionable improvement"),
});

export const ScorecardSchema = z.object({
  clarity: DimensionSchema,
  guidelines: DimensionSchema,
  structure: DimensionSchema,
  examples: DimensionSchema,
});

export const EvaluationSchema = z.object({
  prompt_evaluated: z.string().describe("the user's original prompt, verbatim"),
  scorecard: ScorecardSchema,
  priority_fix: z.string().describe("the single most impactful change to make first"),
  revised_prompt: z.string().describe("the full rewritten prompt, ready to use"),
  suggested_title: z
    .string()
    .describe("a short 3-6 word descriptive title naming what the prompt does"),
  suggested_category: z.string().describe("a single short category for filing"),
  suggested_tags: z.array(z.string()).describe("2-5 short lowercase tags for search"),
});

export type Dimension = z.infer<typeof DimensionSchema>;
export type Scorecard = z.infer<typeof ScorecardSchema>;
export type Evaluation = z.infer<typeof EvaluationSchema>;

export type EvaluationResult = {
  evaluation: Evaluation;
  overall_score: Score;
};
